src/modules/aws_iam@0.0.1/index.ts

```diff
@@ -0,0 +1,72 @@
+import { Role as AWSRole } from '@aws-sdk/client-iam'
+
+import { Role, } from './entity'
+import { AWS,  } from '../../services/gateways/aws'
+import { Context, Crud, Mapper, Module, } from '../interfaces'
+import * as metadata from './module.json'
+
+export const AwsIamModule: Module = new Module({
+  ...metadata,
+  utils: {
+    roleMapper: async (role: AWSRole, ctx: Context) => {
+      const out = new Role();
+      out.arn = role.Arn;
+      out.name = role.RoleName ?? '';
+      out.policyDocument = role.AssumeRolePolicyDocument ?? '';
+      const client = await ctx.getAwsClient() as AWS;
+      out.policyArns = await client.getRolePolicyArns(out.name);
+      return out;
+    },
+  },
+  mappers: {
+    role: new Mapper<Role>({
+      entity: Role,
+      equals: (a: Role, b: Role) => Object.is(a.name, b.name) &&
+        Object.is(a.arn, b.arn) &&
+        Object.is(a.policyDocument, b.policyDocument) &&
+        Object.is(a.policyArns?.length, b.policyArns?.length) &&
+        a.policyArns?.every(as => !!b.policyArns?.find(bs => Object.is(as, bs))),
+      source: 'db',
+      cloud: new Crud({
+        create: async (es: Role[], ctx: Context) => {
+          const client = await ctx.getAwsClient() as AWS;
+          for (const role of es) {
+              const roleArn = await client.newRole(
+                role.name,
+                role.policyDocument,
+                role.policyArns,
+              );
+              if (!roleArn) { // then who?
+                throw new Error('should not be possible');
+              }
+              role.arn = roleArn;
+              await AwsIamModule.mappers.role.db.update(role, ctx);
+          }
+        },
+        read: async (ctx: Context, ids?: string[]) => {
+          const client = await ctx.getAwsClient() as AWS;
+          const roles = Array.isArray(ids) ?
+            await Promise.all(ids.map(id => client.getRole(id))) :
+            (await client.getRoles()) ?? [];
+          return await Promise.all(roles
+            .map(r => AwsIamModule.utils.roleMapper(r, ctx))
+          );
+        },
+        update: async (es: Role[], ctx: Context) => {
```

Comment:
This update is only a replace, so can you add the `updateOrReplace` property to always return replace?

Reply:
Let me look at the module interface

Reply:
I think this is also true for ec2, so let me change it there too